### DISCUSSION TITLE
MIQT Compile benchmarks game

### DISCUSSION THREAD

**Discussion by mappu** [mappu/miqt#65] [General]
The first time you use MIQT, it must be compiled. The first build can be slow, but after this, any `go build` takes only a few seconds.

To assist #8 it would be good to collect numbers from different people about how fast MIQT's first compilation is. Please include CPU, Go version, and MIQT Git tag.

<details>

<summary>🖥️ Example about how to do a clean build and collect all information</summary>

```bash
# Find CPU
$ lscpu | grep -Fi 'model name'
Model name:                           Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz

# Go version
$ go version
go version go1.19.8 linux/amd64

# Git checkout a well-known state
$ git checkout v0.6.0
$ git show --oneline -s
99a29d0c (HEAD, tag: v0.6.0) doc/CHANGELOG: summary for v0.6.0

# Clear GOCACHE for a clean build
$ go clean -cache
$ du -hs "$(go env GOCACHE)"

# Time build (e.g. Qt5 library package)
$ cd qt
$ /usr/bin/time go build -ldflags '-s -w'
532.26user 45.64system 8:50.40elapsed 108%CPU (0avgtext+0avgdata 4878200maxresident)k
713624inputs+4829920outputs (913major+19294096minor)pagefaults 0swaps
```

</details>

**Comment by mappu** [comment]
My results:

|CPU|OS|Go|C/C++|MIQT|Package|Duration|Notes
|---|---|---|---|---|---|---|---
|AMD Ryzen 5 7600 6-Core|Debian 12|1.19.8|GCC 12.2|v0.6.0|qt5|5m 09s
|Intel(R) Core(TM) i7-7700K|Debian 12|1.19.8|GCC 12.2|v0.6.0|qt5|8m 50s
|Neoverse-N1 4-Core|Debian 12|1.19.8|GCC 12.2|v0.6.0|qt5|13m 53s
|Neoverse-N1 4-Core|Debian 12|1.19.8|GCC 12.2 -O0|v0.6.0|qt5|11m 14s|A
|Neoverse-N1 4-Core|Debian 12|1.23.3|GCC 12.2|v0.6.0|qt5|14m 03s|B
|Neoverse-N1 4-Core|Debian 12|1.24-devel|GCC 12.2|v0.6.0|qt5|13m 54s|C
|Neoverse-N1 4-Core|Debian 12|CL 581336|GCC 12.2|v0.6.0|qt5|14m 01s|D
|Neoverse-N1 4-Core|Debian 12|CL 579815|GCC 12.2|v0.6.0|qt5|7m 50s|E
|Github Actions run 2024-11|Debian 12 (Docker)|1.19.8|GCC 12.2|v0.6.0|qt5|9m 22s
|Github Actions run 2024-11|Debian 12 (Docker)|1.19.8|GCC 12.2|v0.6.0|**qt6**|11m 39s
|Github Actions run 2024-11|Fedora 40 (Docker)|1.22|GCC|v0.6.0|**qt6**|11m 55s
|Github Actions run 2024-12|Debian 12 (Docker)|1.19.8|GCC 12.2|v0.7.0+|qt5|13m 24s
|Github Actions run 2024-12|Fedora 41 (Docker)|1.23|GCC|v0.7.0+|**qt6**|17m 6s
|AMD Ryzen 5 7600 2-Core|Win 11 24H2 (VM)|1.22.5|GCC 14.2|v0.7.0+|**qt6**|46m 42s|F
|AMD Ryzen 5 7600 2-Core|Win 11 24H2 (VM)|1.23.1|Clang 18.1|v0.7.0+|**qt6**|32m 34s|G
|AMD Ryzen 5 7600 6-Core|Debian 12|1.19.8|GCC 12.2|v0.7.0+|**qt6**|8m 49s
|AMD Ryzen 5 7600 6-Core|Debian 12|1.19.8|GCC 12.2 -O0|v0.7.0+|**qt6**|6m 41s
|AMD Ryzen 5 7600 6-Core|Debian 12|1.19.8|Clang 14|v0.7.0+|**qt6**|6m 14s
|AMD Ryzen 5 7600 6-Core|Debian 12|1.19.8|Clang 14 -O0|v0.7.0+|**qt6**|5m 56s
|AMD Ryzen 5 7600 2-Core|Debian 12|1.19.8|Clang 14 -O0|v0.7.0+|**qt6**|6m 02s|H
|AMD Ryzen 5 7600 4-Core|FreeBSD 14.2 (VM)|1.21|Clang|v0.7.1|**qt6**|8m 08s

- A. 1.19.8 linux/arm64 with `CGO_CXXFLAGS='-O0' CGO_CFLAGS='-O0'`. This reduced the build duration from 833s to 674s (1.24x faster)
- B. Go1.23 seems to have about the same performance as Go1.19.
- C. `go version devel go1.24-e67c0f0 Fri Nov 8 21:22:55 2024 +0000 linux/arm64 (tip)` (Today's gotip)
- D. `go version devel go1.24-d555358bf8 Tue Sep 3 15:39:35 2024 +0000 linux/arm64` (CL 581336)
- E. `go version devel go1.23-b0a2790fb08 Wed Apr 17 18:48:29 2024 +0100 linux/arm64` (CL 579815). This was the best result on this CPU.
- F. MSYS2 UCRT64, GCC 14.1, Go from MSYS2, Windows Defender enabled but with C:\msys2\ excluded. It is likely that low VM RAM caused swap thrashing
- G. MSYS2 CLANG64, Clang 18.1, Go from go.dev, Windows Defender enabled but with C:\msys2\ excluded. It is likely that low VM RAM caused swap thrashing
- H. Core limited with `taskset -c 0,5`

---

**Summary:**

- Go versions between 1.19 - 1.24 makes no difference in speed, only CL 579815 helps
- Core count (2 vs 6) makes no difference in speed
- Linux builds much faster than Windows
- Qt5 seems to build faster than Qt6
- Clang seems to build faster than GCC
- Building with `-O0` is faster than building with Go's default `-O2`

**Comment by infradragon** [comment]
13th Gen Intel(R) Core(TM) i5-1340P
Arch Linux, CachyOS v3 repositories and kernel
go 1.24.4
miqt v0.11.0 8962d67 (qt6)
`go build -ldflags '-s -w'  1014.23s user 41.90s system 107% cpu 16:20.21 total`
seems to be only using a single core

**Comment by matloob** [comment]
Hello! We're planning to work on fixing golang/go#9887 for Go 1.26, which should improve build times on multicore CPUs. We wanted to check if y'all would be interested in testing out our changes once they are ready.

Here's a "baseline" result. I'll try to post updates with our in progress CLs once they show an improvement.

AMD Ryzen 9 9950X3D 16-Core Processor
go version go1.26-devel_8282b72d62 Wed Aug 6 14:34:39 2025 -0700 linux/amd64 (built from tip yesterday)
miqt v0.11.0 8962d67 (qt6)
7m 38s

**Reply by rcalixte** [reply]
@matloob This looks great! Do you happen to know if dead code elimination and/or some ability to configure/select specific portions of large Cgo dependencies is on the roadmap?

Your compilation time is still over 7 minutes because you are compiling the entire library when only a portion is usually needed per application. If a developer could tune the specific subset that was needed, I suspect the parallel build times could be under half a minute. In addition, the resulting binary is very large, largely due to dead code not being eliminated. These are issues that affect [Wails](https://github.com/wailsapp/wails), particularly if there is a desire to adopt a Qt backend via this library without having to resort to complicated vendoring.

Apologies if this isn't the right place or time! It's great to see this getting attention.

**Reply by mappu** [reply]
I'm very excited to see improvements and i'm happy to test CLs. Thanks for taking a look!

Over-compiling unused parts of Qt definitely increases the number of translation units that have to be compiled and would be great to improve. But I assume it's out of scope. The most important first step here must be, to get parallel compilation of translation units.

Dead code elimination in C++ happens at link-time, and since the cached `go build` times are pretty fast (e.g. 10s vs 10m) and still have to perform linking, the linking must not be a present cause of build-time performance issues unless you start swap thrashing. It is definitely a problem that it doesn't seem to be fully working for us, though. See also https://github.com/mappu/miqt/issues/147 .

**Reply by matloob** [reply]
Yes, reducing over-compiling of unused parts of cgo dependencies is out of scope for this project. Though I'm hoping that compiling the individual C/C++/etc. files in the package will provide a significant performance improvement.

Here's a visualization of what this project will do. The go command has a `--debug-trace <file>` flag that will output a trace in chrome trace viewer format to the given file, how individual parts of package and module loading and action execution are scheduled on to the GOMAXPROCS goroutines we start by default. The traces can be viewed using the `https://ui.perfetto.dev` viewer and opening the files produced.

Here's what the build currently looks like:

<img width="1730" height="1299" alt="Screenshot 2025-08-10 at 08-42-03 miqt trace (1 MB) - Perfetto UI" src="https://github.com/user-attachments/assets/7f0c218f-d368-471f-9f25-c4a28fb579dc" />

First we run cgo, the first blue bar, then compile all the files sequentially (all the small multi colored bars after it), and finally do the normal build steps (the area after the bottom set of bars end). Running cgo takes about 2m 10s, and running each of the actions takes about 4m 49s.

Here's a view of what our goal is (produced using a proof-of-concept prototype I'm working on). Now the length of the trace represents about 1 minute rather than 7m38s.

<img width="1728" height="1301" alt="Screenshot 2025-08-10 at 08-41-39 miqt trace par (1 MB) - Perfetto UI" src="https://github.com/user-attachments/assets/483a13fb-5530-4864-8244-1ea1d771bec3" />

We've parallelized some of the c compilations done by cgo (this is really experimental and may not be possible in the end, but I hope we can make it work) taking the cgo step to about 18s (it's a single blue bar because it's a single action to invoke cgo by the go command), and the actions for compiling the c files (which are now saturating out action goroutines) which now takes about 20s on my machine. The go compiler invocation takes about the same time (though it may be possible to make this go faster by increasing the concurrency argument to the compiler, which is set to a default of 4).

Of course we may not be able to realize all these gains in the final version, and the improvement won't be quite as big on machines with fewer cores, but that's the idea of what we're trying.

**Reply by matloob** [reply]
@mappu Hello! I have a stack of changes that's ready for initial testing. Please check out https://golang.org/cl/699435. I have instructions in the change description for how to try it out. Please let me know if you have any questions.

**Reply by mappu** [reply]
Thanks! It helps a lot - a sample build drops from 6m20s to 1m40s. :tada: 

I have some issues with the `#cgo` directives, i've posted more information in the other thread. But i'm sure that's easy to solve. Thank you so much, this is a really fantastic improvement,

**Reply by matloob** [reply]
@mappu The cgo changes have all been submitted into tip. There may be some ways that we can improve things further, but I think we've made the bulk of the improvements and there are perhaps diminishing returns after.

But just to document some of the ideas:
- We could sort files by size and start processing them when they're ready. I have a prototype of that here: https://go-review.googlesource.com/c/go/+/704276
- We could cache the cgo and c compile outputs separately from the go object, allowing us to break dependency edges from those actions to builds of package dependencies.
- We could compile the dynimport file produced by cgo separately from the rest of the package allowing us to break dependency edges to builds of dependent package.

To connect these to the second "after" trace visualized above:, the first idea would try to shorten the pink box, the second would try to allow the pink box to run in parallel with builds of other packages, and the third would try to allow the purple box to run in parallel with builds of other packages.

**Reply by matloob** [reply]
We've made some more changes in the Go 1.28 tree which now decouple the build action from the cgo compile actions, so first-time builds should be even faster! At this point we're pretty much just limited on cmd/cgo.
